freedom/reco/crs_reco.py

```diff
@@ -32,43 +35,69 @@
 from freedom.reco import summary_df
 from freedom.reco import prefit
 
+#load time residual spline function
+from scipy.interpolate import UnivariateSpline
+from freedom.utils.time_residual import TimeResidual
+with open(pkg_resources.resource_filename('freedom', 'resources/prior/oscNext_time_residual_prior.pkl'), 'rb') as F:
+    ftr = pickle.load(F)
+ftr = UnivariateSpline._from_tck(ftr)
+ftr.ext = 1
 
-def get_batch_closure(client, event, out_of_bounds):
+
+def get_batch_closure(clients, event, out_of_bounds, Tprior=True):
     """returns LLH batch eval closure for this event. The closure is a function of one argument: the hypotheses to evaluate"""
     hit_data = event["hit_data"]
     evt_data = event["evt_data"]
-
+    
     def eval_llh(params):
-        llhs = np.atleast_1d(client.eval_llh(hit_data, evt_data, params))
+        llhs = 0
+        for i in range(len(clients)):
+            llhs += np.atleast_1d(clients[i].eval_llh(hit_data[i], evt_data[i], params))
+            
+            if Tprior:
+                tresiduals = TimeResidual(np.tile(hit_data[i], (len(llhs),1)).astype(np.float32), 
+                                          np.repeat(params, len(hit_data[i]), axis=0).astype(np.float32))
+                tresiduals = np.log(np.clip(ftr(tresiduals), 1e-10, 1))
+                llhs -= np.sum(tresiduals.reshape((len(llhs), len(hit_data[i]))), axis=1)
 
         return bounds.invalid_replace(llhs, params, out_of_bounds)
-
+    
     return eval_llh
 
 
 def batch_crs_fit(
     event,
-    client,
+    clients,
     rng,
     init_range,
     search_limits,
     n_live_points,
     bounds_check_type="cube",
+    Tprior=True,
     **sph_opt_kwargs,
 ):
 
     out_of_bounds = bounds.get_out_of_bounds_func(search_limits, bounds_check_type)
 
-    eval_llh = get_batch_closure(client, event, out_of_bounds)
+    eval_llh = get_batch_closure(clients, event, out_of_bounds, Tprior=Tprior)
 
     n_params = len(init_range)
 
-    box_limits = prefit.initial_box(event["hit_data"], init_range, n_params=n_params)
+    a = np.array([])
```

Comment:
what are the cryptic `a`s and `e`s?

Reply:
This code was added for ICU reco (where you can have 0 hits for a module type). a just means array and e event, I will add some comments